adapters.go

```diff
@@ -194,136 +186,80 @@ func (a *whisperAdapter) handleRetrievedMessages(messages []*whisper.ReceivedMes
+			statusMessages, err := a.handleMessages(shhMessage, false)
 			if err != nil {
-				logger.Warn("failed to handle an encryption message", zap.Error(err), zap.Binary("messageID", shhMessage.Hash))
+				logger.Info("failed to decode messages", zap.Error(err))
+				continue
 			}
-			result[chat] = append(result[chat], shhMessage)
-		}
-	}
-
-	return result, nil
-}
-
-// DEPRECATED
-func (a *whisperAdapter) RetrieveRaw(filterID string) ([]*whisper.Message, error) {
-	messages, err := a.transport.RetrieveRaw(filterID)
-	if err != nil {
-		return nil, err
-	}
-
-	logger := a.logger.With(zap.String("site", "RetrieveRaw"))
 
-	var result []*whisper.Message
+			result[chat] = append(result[chat], statusMessages...)
 
-	for _, message := range messages {
-		shhMessage := whisper.ToWhisperMessage(message)
-		err := a.handleMessageEncryption(context.Background(), shhMessage)
-		if err != nil {
-			logger.Warn("failed to handle an encryption message", zap.Error(err), zap.Binary("messageID", shhMessage.Hash))
 		}
-		result = append(result, shhMessage)
 	}
 
 	return result, nil
 }
 
-func (a *whisperAdapter) addDatasyncPacket(publicKey *ecdsa.PublicKey, datasyncMessage datasyncproto.Payload) {
-	packet := datasynctransport.Packet{
-		Sender:  datasyncpeer.PublicKeyToPeerID(*publicKey),
-		Payload: datasyncMessage,
-	}
-	a.datasync.AddPacket(packet)
-}
-
-// handleDecodedMessages expects an unencrypted whisper message as it's argument.
-// 1) Check if it's a datasync message
-// 2) If it's a datasync message it will handle each transmitted message separately
-// 3) if it's not a datasync message a single message will be handled
-// Fingerprinting datasync messages is a bit tricky, as they might unmarshal fine
-// because they are wrapped messages, so we need to check whether anything is
-// poulated in the fields.
-func (a *whisperAdapter) handleDecodedMessages(message *whisper.Message) ([]*protocol.StatusMessage, error) {
-	var decodedMessages []*protocol.StatusMessage
-
+// handleMessages expects a whisper message as input, and it will go through
+// a series of transformations until the message is parsed into an application
+// layer message, or in case of Raw methods, the processing stops at the layer
+// before
+func (a *whisperAdapter) handleMessages(shhMessage *whisper.Message, applicationLayer bool) ([]*protocol.StatusMessage, error) {
 	logger := a.logger.With(zap.String("site", "decodeMessages"))
-	hlogger := logger.With(zap.Binary("hash", message.Hash))
+	hlogger := logger.With(zap.Binary("hash", shhMessage.Hash))
+	var statusMessage protocol.StatusMessage
 
-	publicKey, err := crypto.UnmarshalPubkey(message.Sig)
+	err := statusMessage.HandleTransport(shhMessage)
 	if err != nil {
-		return nil, err
+		hlogger.Error("failed to handle transport layer message", zap.Error(err))
 	}
 
-	// Try first with datasync
-	datasyncMessage, err := protocol.UnwrapDatasync(message.Payload)
+	err = a.handleEncryptionLayer(context.Background(), &statusMessage)
```

Comment:
We have a bit of mix of `handle*` messages on the adapter and `Handle*` on the `StatusMessage`. 

To clean up this flow, we could have `a.handle*()` for each layer which would also took care of logging errors.